### ISSUE TITLE
Support Remote Audit Log Server Configuration

### ISSUE THREAD

**Issue by tharikaGitHub** [wso2/api-manager#1565]
### Description

Currently we do not have the capability to configure the Audit Server URL configuration through the UI. Apart from adding this capability, we need to generate audit logs for the modification of the configuration.

### Affected Component

APIM

### Version

4.2.0

**Comment by tharikaGitHub** [comment]
**Update** : 

According to [1] there are few drawbacks when trying to use the programmatic way in modifying the log4j2 configuration. The following issue as mentioned in the previous comment [2] occurs when trying to use these methods. 

`java.lang.ClassCastException: org.ops4j.pax.logging.log4jv2.Log4jv2LoggerContext cannot be cast to org.apache.logging.log4j.core.LoggerContext`

Hence I have used a different method of modifying the configuration file. There is no change to the configurations even when the server is restarted. Hence we are good to go with the second approach. 

I have done the required code changes to make the functionality work as expected. I'm currently working on enhancing the existing code.

The UI for this feature looks as below as of now.

<img width="1512" alt="Screen Shot 2023-03-21 at 7 04 32 PM" src="https://user-images.githubusercontent.com/8557410/226622160-320f5215-aaa0-4034-b17b-c728c1e5db0f.png">
After few more things are clarified we are good to go with further changes as required.

[1] https://logging.apache.org/log4j/2.x/manual/customconfig.html#programmatically-modifying-the-current-configuration-after-initi
[2] https://github.com/wso2/api-manager/issues/1565#issuecomment-1473799111

**Comment by tharikaGitHub** [comment]
**Update** : 

I have completed the enhancement fixes and refactoring. A draft PR[1] is opened with the changes to the main repo. The UI is as follows.
<img width="1509" alt="Screen Shot 2023-03-22 at 8 45 54 PM" src="https://user-images.githubusercontent.com/8557410/226951584-003938da-2555-4bd6-85f4-318e4b6b02c9.png">

Also an audit log is added to indicate successful configuration changes to log4j2.properties file when configuring the remote server url.

Now this fix should be included in the relevant products by way of feature installation + testing.

[1] https://github.com/wso2/carbon-commons/pull/466

**Comment by tharikaGitHub** [comment]
**Update** : 

Added the above feature to the carbon-commons branch relevant to Identity Server product and added the feature installation to product-is. The draft PRs can be found in [1] and [2]

Now we need to add the additional fields as mentioned in comment [3] and modify the implementation accordingly. 

[1] https://github.com/wso2/carbon-commons/pull/467
[2] https://github.com/wso2/product-is/pull/15651
[3] https://github.com/wso2/api-manager/issues/1565#issuecomment-1481382339

**Comment by tharikaGitHub** [comment]
It was required that we support the configuration of Audit Logs, API Logs and Carbon Logs as well through the UI. Hence the UI was modified to include that aspect as follows. I'm currently working on its implementation.

<img width="840" alt="Screen Shot 2023-03-31 at 6 42 29 PM" src="https://user-images.githubusercontent.com/8557410/229129450-bf92e86d-c57e-4866-81c1-c129480bd56a.png">

**Comment by tharikaGitHub** [comment]
**Update** : 

I have added the functionality to configure the 3 required types of logs Audit Logs, API Logs and Carbon Logs as follows. The changes are added to the draft PR [1]. The same fixes should go to the carbon-commons branch used by Identity Server Product with slight changes on the type of logs. (API Logs are not available in Identity Server).

<img width="1512" alt="Screen Shot 2023-04-03 at 5 01 35 PM" src="https://user-images.githubusercontent.com/8557410/229497234-3af4ba0c-5d0f-44b6-89dc-250a59aeb895.png">

A separate section called **Advanced Configurations** was introduced to add all the extra optional information related to the remote server url being configured. See the image below. Currently it has the config to specify the connection timeout. We need to add the security configurations as well here.

<img width="1512" alt="Screen Shot 2023-04-03 at 5 02 25 PM" src="https://user-images.githubusercontent.com/8557410/229497577-b23a2905-177f-45ff-bb79-655a5ef80927.png">

Currently there is no option to revert back to the default configs from the UI. This capability should also be introduced for being more user friendly.

[1] https://github.com/wso2/carbon-commons/pull/466

**Comment by tharikaGitHub** [comment]
**Update** : 

I have added the changes required for Identity Server Product as well to the carbon-commons repo with draft PR [1]. The UI will look like bellow without the API Logs option.

<img width="1512" alt="Screen Shot 2023-04-04 at 9 31 05 AM" src="https://user-images.githubusercontent.com/8557410/229684756-93349b00-98e5-43c9-bbdf-c06d2db96793.png">

[1] https://github.com/wso2/carbon-commons/pull/467

**Comment by tharikaGitHub** [comment]
**Update** : 

I have added the option to revert back to the default configs from the UI to both the APIM and IS products. The code changes are added to the draft PRs [1] and [2].

Apart from the above I have renamed the modules to be consistent with the feature and the changes are added to the draft PRs [1], [2], [3] and [4].

[1] https://github.com/wso2/carbon-commons/pull/466
[2] https://github.com/wso2/carbon-commons/pull/467
[3] https://github.com/wso2/product-apim/pull/13188
[4] https://github.com/wso2/product-is/pull/15651

**Comment by tharikaGitHub** [comment]
**Update** : 

I have added the capability to configure a remote server which requires **Basic Auth** to be supported. The UI for APIM is as follows.
<img width="1512" alt="Screen Shot 2023-04-13 at 12 00 04 PM" src="https://user-images.githubusercontent.com/8557410/231673560-61bb2e5a-b7f8-4819-bc38-a602af8a42c7.png">

The code changes are added in draft PR [1].

I need to add this capability to Identity Server product as well. We also need to look into how secure vault can be used here.
The next step after that would be to support SSL configurations for https remote servers.

[1] https://github.com/wso2/carbon-commons/pull/466

**Comment by tharikaGitHub** [comment]
**Update** : 

I have added the support to configure SSL configuration for the Remote Server in APIM. The latest changes are added to the draft PR [1]. The UI is as follows.

<img width="1512" alt="Screen Shot 2023-04-15 at 5 55 11 AM" src="https://user-images.githubusercontent.com/8557410/232175374-2182b3a4-be77-45f4-af12-8415cf1fc0cf.png">

**_Next action items_** : 

- [ ] Add the functionality to the Identity Server Product
- [ ] Check how secure vault can be used to protect the sensitive information in the log4j2.properties file
- [ ] Test the SSL configuration functionality with an https server

[1] https://github.com/wso2/carbon-commons/pull/466

**Comment by tharikaGitHub** [comment]
**Update** : 

- The code changes are added to the APIM and IS products with the PRs [1], [2], [3] and [4]. Those PRs are added to the "ready for review" state.
- As we are adding passwords to the appender configuration in the log4j2.properties file, it is good to secure them using secure vault. But it does not work with the third party custom appenders. In this case HTTPAppender. The reason is, third-party custom Log4j2 appenders do not have the same wso2 secure-vault implementation to resolve the encrypted passwords. Therefore, whenever the encrypted passwords are used along with these custom Log4j2 appenders, we cannot expect them to get resolved OOB. Instead we can write a custom Log4j2 appender by extending the existing HTTPAppender in Log4j2 adding the secure vault capability or else ask the user to add the passwords as environment variables. This has been the previous recommendation. We need to come to a conclusion on this.
- A code review will be scheduled next week to finalize the code.

[1] https://github.com/wso2/carbon-commons/pull/466
[2] https://github.com/wso2/carbon-commons/pull/467
[3] https://github.com/wso2/product-apim/pull/13188
[4] https://github.com/wso2/product-is/pull/15651

**Comment by tharikaGitHub** [comment]
**Update** :

We had a code review on the feature today. Please find the call notes below.

- Check if the audit log that has been added will work with the forget me tool.
- Check if the existing implementations for secure vault in log4j2.properties file would work with this feature. If not we need to write a custom appender to do the resolving.
- Provide a Reset option for the user.
- Check for UI security when accessing the user input parameters

**Comment by lasanthaS** [comment]
Hi all,

I started working on writing a Custom HTTP log appender that supports secure-vault. However, I encountered an issue which the Log4j2 doesn't pick up the log appender. I am currently looking into that issue.

Thanks,
Lasantha

**Comment by lasanthaS** [comment]
Hi all,

I have raised PR to update the `forgetme` tool to replace the audit logs that has been added with this feature. The PR can be found at https://github.com/wso2/identity-anonymization-tool/pull/165. 

Now I am working on providing the reset option for the user, and the UI-related changes are done. 

Thanks,
Lasantha

**Comment by lasanthaS** [comment]
Hi all,

I have added the support to reset the remote audit server log configuration via the option in the UI. This fix is delivered via the PR https://github.com/wso2/carbon-commons/pull/471 (commit https://github.com/wso2/carbon-commons/pull/471/commits/78e171194905befd578c9f5697fc7bc13934251b).

Regarding the task "Check for UI security when accessing the user input parameters", there are no places where the user input parameters are stored and displayed on UI, instead of applying them into the `log4j2.properties` file. Therefore, it is not necessary to encode them as encoding these values will break the log4j2 configurations.

**Pending Tasks:**

- [x] Check if the audit log that has been added will work with the forget me tool.
- [ ] Check if the existing implementations for secure vault in log4j2.properties file would work with this feature. If not we need to write a custom appender to do the resolving.
- [ ] Provide a Reset option for the user.
- [ ] Check for UI security when accessing the user input parameters

Thanks,
Lasantha

**Comment by lasanthaS** [comment]
Hi all,

I created a custom HTTP log appender, which publishes the log messages to a remote HTTP server by extending `org.apache.logging.log4j.core.appender.AbstractAppender` class. As per the current implementation, the user can pass the username and the password via appender attributes, then the appender builds the basic auth header from the given credentials and appends it to each request.  

However, after adding support to the secure vault I faced an issue in which the secure vault is not initialized when the appender is loaded. Because of that, we cannot decrypt the password on the appender initialization phase.  In this case, we can't change OSGi load order, as the logging framework is one of the key components which initially starts (priority 0). Therefore, we can't amend the load order as well.

As a workaround for this, we thought of introducing a queue to hold the log events, then have a separate thread that starts consuming the log events once the secret resolver is initialized and publishes those events to the remote server.

I am currently implementing that feature and will update this once the implementation is done.

Thanks,
Lasantha

**Comment by lasanthaS** [comment]
Hi all,

I was able to complete the implementation of the HTTP log appender which supports secure-vault. 

Within the log appender, I introduced a `BlockingQueue` which buffers the log events and had a separate `ScheduledExecutorService` to read the events from the queue and publish it to the HTTP endpoint once the secure-vault is initialized properly.

However, there I had an issue identifying the exact point when the secure-vault is initialized properly. To cater to that, I tried adding an OSGi service component that waits from the `ConfigurationContextService` class object. But again adding that OSGi component to the log appender component was not possible as the log appender component had the `Fragment-Host` declared.

```xml
<Fragment-Host>org.ops4j.pax.logging.pax-logging-log4j2</Fragment-Host>
```

As a workaround for it, I introduced another component called `org.wso2.carbon.log.appender.service` which hosts the OSGi service component and sets a flag in the log appender component once the `ConfigurationContextService` is enabled. This fixed the issue.

Now I am testing the feature and need to include these components within the feature itself. I will raise the PR once this is done.

Thanks,
Lasantha

**Comment by lasanthaS** [comment]
Hi all,

Updated the PR[1] with the secure vault fixes and the feature to install the http appender.

[1] https://github.com/wso2/carbon-commons/pull/471/

**Comment by lasanthaS** [comment]
Hi all,

Please find the PRs related to secure vault fixes and the feature to install the HTTP appender below.

APIM:
- https://github.com/wso2/carbon-commons/pull/471
- https://github.com/wso2/product-apim/pull/13220

IS:
- https://github.com/wso2/carbon-commons/pull/474
- https://github.com/wso2/product-is/pull/16094

Thanks,
Lasantha

**Comment by tharikaGitHub** [comment]
Closing this issue as this feature is already available in APIM 4.3.0-Alpha.
